src/Misc/DegreeLocalization.jl

```diff
@@ -0,0 +1,396 @@
+
+export KInftyRing, KInftyElem, function_field
+
+###############################################################################
+#
+#  Declaration types
+#  KInftyRing / KInftyElem 
+#
+###############################################################################
+
+mutable struct KInftyRing{T <: FieldElement} <: Hecke.Ring
+   K::Generic.RationalFunctionField{T}
+
+   function KInftyRing{T}(K::Generic.RationalFunctionField{T}, cached::Bool) where T <: FieldElement
+      return AbstractAlgebra.get_cached!(KInftyID, K, cached) do
+         new{T}(K)
+      end::KInftyRing{T}
+   end
+end
+
+const KInftyID = Dict{Generic.RationalFunctionField, Hecke.Ring}()
+
+mutable struct KInftyElem{T <: FieldElement} <: Hecke.RingElem
+   d::Generic.Rat{T}
+   parent::KInftyRing{T}
+end
+
+###############################################################################
+#
+#   Data type and parent object methods
+#
+###############################################################################
+
+elem_type(::Type{KInftyRing{T}}) where T <: FieldElement = KInftyElem{T}
+
+parent_type(::Type{KInftyElem{T}}) where T <: FieldElement = KInftyRing{T}
+
+function function_field(R::KInftyRing{T}) where T <: FieldElement
+   return R.K::Generic.RationalFunctionField{T}
+end
+
+parent(a::KInftyElem{T}) where T <: FieldElement = a.parent
+
+function check_parent(a::KInftyElem{T}, b::KInftyElem{T})  where T <: FieldElement
+   parent(a) != parent(b) && error("Parent objects do not match")
+end
+
+###############################################################################
+#
+#   Basic manipulation
+#
+###############################################################################
+
+data(a::KInftyElem{T}) where T <: FieldElement = a.d::Generic.Rat{T}
+
+function numerator(a::KInftyElem{T}, canonicalise::Bool=true) where T <: FieldElement
+   return numerator(data(a), canonicalise)
+end
+
+function denominator(a::KInftyElem{T}, canonicalise::Bool=true) where T <: FieldElement
+   return denominator(data(a), canonicalise)
+end
+
+degree(a::KInftyElem) = degree(numerator(a, false)) - degree(denominator(a, false))
+
+valuation(a::KInftyElem) = -degree(a)
+
+zero(K::KInftyRing{T}) where T <: FieldElement = K(0)
+
+one(K::KInftyRing{T}) where T <: FieldElement = K(1)
+
+iszero(a::KInftyElem{T}) where T <: FieldElement = iszero(data(a))
+
+isone(a::KInftyElem{T}) where T <: FieldElement = isone(data(a))
+
+function isunit(a::KInftyElem{T}) where T <: FieldElement
+   return degree(numerator(data(a), false)) ==
+                                            degree(denominator(data(a), false))
+end
+
+function in(a::Generic.Rat{T}, R::KInftyRing{T}) where T <: FieldElement
+   if parent(a) != function_field(R)
+      return false
+   end
+   return degree(numerator(a, false)) <= degree(denominator(a, false))
+end
+
+function deepcopy_internal(a::KInftyElem{T}, dict::IdDict) where T <: FieldElement
+   parent(a)(deepcopy(data(a)))
```

Comment:
Probably should be passed `dict`?

Reply:
Are you sure? I'm having trouble finding any other examples of where we do so. I haven't the slightest idea what it does by the way.